### PR TITLE
Implementation of new editor completion API

### PR DESCRIPTION
**Pull request by dpoeschl**
Replaced by review https://github.com/dotnet/roslyn/pull/25770

<details><summary>Ask Mode template not completed</summary>

### Customer scenario

What does the customer do to get into this situation, and why do we think this
is common enough to address for this release.  (Granted, sometimes this will be
obvious "Open project, VS crashes" but in general, I need to understand how
common a scenario is)

### Bugs this fixes

(either VSO or GitHub links)

### Workarounds, if any

Also, why we think they are insufficient for RC vs. RC2, RC3, or RTW

### Risk

This is generally a measure our how central the affected code is to adjacent
scenarios and thus how likely your fix is to destabilize a broader area of code

### Performance impact

(with a brief justification for that assessment (e.g. "Low perf impact because no extra allocations/no complexity changes" vs. "Low")

### Root cause analysis

How did we miss it?  What tests are we adding to guard against it in the future?

### How was the bug found?

(E.g. customer reported it vs. ad hoc testing)

### Test documentation updated?

If this is a new non-compiler feature or a significant improvement to an existing feature, update https://github.com/dotnet/roslyn/wiki/Manual-Testing once you know which release it is targeting.

</details>

### CODE REVIEW

**Comment by heejaechang** [comment]
@jinujoseph @dpoeschl can we have share knowledge thing next Tuesday on this?

File: src/EditorFeatures/Core/Implementation/IntelliSense/EditorCompletionImplementation/CompletionItemSource.cs [hunk 1]
```diff
@@ -0,0 +1,295 @@
...
+{
+    internal class CSharpCompletionItemSource : AbstractCompletionItemSource
+    {
+        internal override bool PassEnterThroughToBuffer() => false;
+    }
+
+    internal class VisualBasicCompletionItemSource : AbstractCompletionItemSource
+    {
+        internal override bool PassEnterThroughToBuffer() => true;
+    }
```
Comment (CyrusNajmabadi) [inline src/EditorFeatures/Core/Implementation/IntelliSense/EditorCompletionImplementation/CompletionItemSource.cs]: feels weird that these types are in a common location.

File: src/EditorFeatures/Core/Implementation/IntelliSense/EditorCompletionImplementation/CompletionItemSourceProvider.cs [hunk 2]
```diff
@@ -0,0 +1,39 @@
...
+
+using System.ComponentModel.Composition;
+using Microsoft.VisualStudio.Language.Intellisense;
+using Microsoft.VisualStudio.Text.Editor;
+using Microsoft.VisualStudio.Utilities;
+
+namespace Microsoft.CodeAnalysis.Editor.Implementation.IntelliSense.Completion.EditorImplementation
+{
+    [Export(typeof(IAsyncCompletionItemSourceProvider))]
+    [Name("C# Item Source Provider")]
```
Comment (CyrusNajmabadi) [inline src/EditorFeatures/Core/Implementation/IntelliSense/EditorCompletionImplementation/CompletionItemSourceProvider.cs]: seems weird this is in common location.

File: src/EditorFeatures/Core/Implementation/IntelliSense/EditorCompletionImplementation/CompletionItemSourceProvider.cs [hunk 3]
```diff
@@ -0,0 +1,39 @@
...
+    [ContentType(ContentTypeNames.CSharpContentType)]
+    internal class CSharpCompletionItemSourceProvider : IAsyncCompletionItemSourceProvider
+    {
+        CSharpCompletionItemSource _instance;
+
+        IAsyncCompletionItemSource IAsyncCompletionItemSourceProvider.GetOrCreate(ITextView textView)
+        {
+            if (_instance == null)
```
Comment (CyrusNajmabadi) [inline src/EditorFeatures/Core/Implementation/IntelliSense/EditorCompletionImplementation/CompletionItemSourceProvider.cs]: curlies

File: src/EditorFeatures/Core/Implementation/IntelliSense/EditorCompletionImplementation/CompletionItemSourceProvider.cs [hunk 4]
```diff
@@ -0,0 +1,39 @@
...
+using Microsoft.VisualStudio.Utilities;
+
+namespace Microsoft.CodeAnalysis.Editor.Implementation.IntelliSense.Completion.EditorImplementation
+{
+    [Export(typeof(IAsyncCompletionItemSourceProvider))]
+    [Name("C# Item Source Provider")]
+    [ContentType(ContentTypeNames.CSharpContentType)]
+    internal class CSharpCompletionItemSourceProvider : IAsyncCompletionItemSourceProvider
+    {
+        CSharpCompletionItemSource _instance;
```
Comment (CyrusNajmabadi) [inline src/EditorFeatures/Core/Implementation/IntelliSense/EditorCompletionImplementation/CompletionItemSourceProvider.cs]: modifiers.

File: src/EditorFeatures/Core/Implementation/IntelliSense/EditorCompletionImplementation/EditorAsyncCompletionServiceProvider.cs [hunk 5]
```diff
@@ -0,0 +1,36 @@
...
+using System.ComponentModel.Composition;
+using Microsoft.VisualStudio.Language.Intellisense;
+using Microsoft.VisualStudio.Text.Editor;
+using Microsoft.VisualStudio.Utilities;
+using EditorCompletion = Microsoft.VisualStudio.Language.Intellisense;
+
+namespace Microsoft.CodeAnalysis.Editor.Implementation.IntelliSense.Completion.EditorImplementation
+{
+    [Export(typeof(IAsyncCompletionServiceProvider))]
+    [Name("C# and Visual Basic Completion Service Provider")]
```
Comment (CyrusNajmabadi) [inline src/EditorFeatures/Core/Implementation/IntelliSense/EditorCompletionImplementation/EditorAsyncCompletionServiceProvider.cs]: Is the expectation that F#/TS/JS export this as well.  Are they already doing this?

File: src/EditorFeatures/Test2/EditorServicesTest2.vbproj [hunk 6]
```diff
@@ -2,6 +2,7 @@
 <!-- Copyright (c)  Microsoft.  All Rights Reserved.  Licensed under the Apache License, Version 2.0.  See License.txt in the project root for license information. -->
 <Project Sdk="Microsoft.NET.Sdk">
   <PropertyGroup>
+    <NoWarn>40000</NoWarn>
```
Comment (CyrusNajmabadi) [inline src/EditorFeatures/Test2/EditorServicesTest2.vbproj]: what's this for?

File: src/EditorFeatures/TestUtilities/ServicesTestUtilities.csproj [hunk 7]
```diff
@@ -2,6 +2,7 @@
 <!-- Copyright (c)  Microsoft.  All Rights Reserved.  Licensed under the Apache License, Version 2.0.  See License.txt in the project root for license information. -->
 <Project Sdk="Microsoft.NET.Sdk">
   <PropertyGroup>
+    <NoWarn>0618;1591;1573</NoWarn>
```
Comment (CyrusNajmabadi) [inline src/EditorFeatures/TestUtilities/ServicesTestUtilities.csproj]: what are these for?

File: src/Interactive/EditorFeatures/Core/InteractiveEditorFeatures.csproj [hunk 8]
```diff
@@ -2,6 +2,7 @@
 <!-- Copyright (c)  Microsoft.  All Rights Reserved.  Licensed under the Apache License, Version 2.0.  See License.txt in the project root for license information. -->
 <Project Sdk="Microsoft.NET.Sdk">
   <PropertyGroup>
+    <NoWarn>0618;1591</NoWarn>
```
Comment (CyrusNajmabadi) [inline src/Interactive/EditorFeatures/Core/InteractiveEditorFeatures.csproj]: ?

File: src/VisualStudio/IntegrationTest/IntegrationTests/CSharp/CSharpIntelliSense.cs [hunk 9]
```diff
@@ -281,27 +281,27 @@ class Class1
...
-
-            VisualStudio.Editor.SendKeys(
-                VirtualKey.Delete,
-                "aaa",
-                VirtualKey.Tab);
-            var actualText = VisualStudio.Editor.GetText();
-            Assert.Contains("aaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaa = aaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaa", actualText);
-            Assert.True(VisualStudio.Editor.IsCaretOnScreen());
-        }
+//        [Fact, Trait(Traits.Feature, Traits.Features.Completion)]
```
Comment (CyrusNajmabadi) [inline src/VisualStudio/IntegrationTest/IntegrationTests/CSharp/CSharpIntelliSense.cs]: why?

File: src/VisualStudio/IntegrationTest/TestUtilities/OutOfProcess/Editor_OutOfProc.cs [hunk 10]
```diff
@@ -141,8 +141,8 @@ public void DeleteText(string text)
         public void ReplaceText(string oldText, string newText)
             => _editorInProc.ReplaceText(oldText, newText);
 
-        public bool IsCaretOnScreen()
-            => _editorInProc.IsCaretOnScreen();
+        //public bool IsCaretOnScreen()
+        //    => _editorInProc.IsCaretOnScreen();
```
Comment (CyrusNajmabadi) [inline src/VisualStudio/IntegrationTest/TestUtilities/OutOfProcess/Editor_OutOfProc.cs]: why?

**Comment by dpoeschl** [comment]
Closing in favor of https://github.com/dotnet/roslyn/pull/25770

**Pull request closed without merging**